pandas/core/index.py

```diff
@@ -1244,8 +1272,12 @@ def putmask(self, mask, value):
         numpy.ndarray.putmask
         """
         values = self.values.copy()
-        np.putmask(values, mask, value)
-        return self._shallow_copy(values)
+        try:
+            np.putmask(values, mask, value)
```

Comment:
see my very last comment, if you do this:

```
try:
    np.putmask(values, mask, self._convert_for_op(value))
    return self._shallow_copy(values)
except (ValueError, TypeError):
    # coerces to object
    .......
```

I think you can get this to work by only overriding ``_convert_to_value`` for ``DatetimeIndex/PeriodIndex`` (where it would raise if its not compatabile, else return an appropriate value